### DISCUSSION TITLE
MIDI drivers?

### DISCUSSION THREAD

**Discussion by snej** [toitlang/toit#2430] [Drivers]
I’m interested in Toit on ESP32 to interface with synthesizers. There doesn’t appear to be a MIDI driver available (either USB or serial.) 

There’s a USB MIDI driver in the ESP-IDF; how hard would it be to wrap a Toit API around that?

Alternatively, one could implement MIDI in Toit, talking directly to a serial port or a generic USB driver. MIDI is a pretty simple protocol, but I’ve never worked with serial or USB directly before...

**Comment by floitsch** [comment]
Typically we prefer writing drivers in Toit when possible. If midi is a simple protocol then having the code in Toit would make sense, and make it much more flexible. For example you could then develop and test on a normal computer and run the same program (with only tiny modifications) on the ESP32.

Note that Toit doesn't really have any USB support yet, so communication would need to be over the UART (serial port).
